src/internal/easee/auth_test.go

```diff
@@ -0,0 +1,380 @@
+		username                  string
+		password                  string
+		loginStatus               int
+		accessToken               string
+		refreshToken              string
+		saveError                 error
+		errorContains             string
+		notificationManagerCalled int
+	}{
+		{
+			name:          "should return error when login has failed",
+			loginStatus:   http.StatusBadRequest,
+			errorContains: "expected response code to be 200",
+		},
+		{
+			name:          "should return error when storage failed to save",
+			username:      "user",
+			password:      "pwd",
+			loginStatus:   http.StatusOK,
+			accessToken:   "accessToken",
+			refreshToken:  "refreshing",
+			saveError:     errors.New("failed to save to the storage"),
+			errorContains: "failed to save to the storage",
+		},
+		{
+			name:                      "should save tokens to the storage",
+			username:                  "user",
+			password:                  "pwd",
+			loginStatus:               http.StatusOK,
+			accessToken:               "accessToken",
+			refreshToken:              "refreshing",
+			notificationManagerCalled: 1,
+		},
+	}
+
+	for _, val := range testCases {
+		v := val
+		t.Run(v.name, func(t *testing.T) {
+			t.Parallel()
+
+			handler := func(w http.ResponseWriter, r *http.Request) {
+				data := fmt.Sprintf(`{"accessToken":"%s","refreshToken":"%s"}`,
+					v.accessToken,
+					v.refreshToken,
+				)
+				w.WriteHeader(v.loginStatus)
+				_, _ = w.Write([]byte(data))
+			}
+			server := httptest.NewServer(http.HandlerFunc(handler))
+			defer server.Close()
+
+			cfg := config.Config{}
+			storage := mockedstorage.Storage{}
+			storage.On("Model").Return(&cfg)
+			storage.On("Save").Return(v.saveError)
+
+			notificationManager := &NotificationMock{}
+			notificationManager.On("Event", &notification.Event{EventName: notificationEaseeStatusOnline}).Return(nil)
+
+			httpClient := &http.Client{Timeout: 3 * time.Second}
+			client := NewHTTPClient(httpClient, server.URL)
+			auth := authenticator{http: client, cfgSvc: config.NewService(&storage), notificationManager: notificationManager}
+
+			err := auth.Login(v.username, v.password)
+			if err != nil {
+				assert.Contains(t, err.Error(), v.errorContains)
+			} else {
+				assert.Equal(t, v.accessToken, cfg.AccessToken)
+				assert.Equal(t, v.refreshToken, cfg.RefreshToken)
+				assert.Equal(t, statusWorkingProperly, auth.status)
+				notificationManager.AssertNumberOfCalls(t, "Event", v.notificationManagerCalled)
+			}
+		})
+	}
+}
+
+func TestAccessToken(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		name          string
+		credentials   config.Credentials
+		authStatus    connectivityStatus
+		refreshStatus int
+		accessToken   string
+		refreshToken  string
+		saveError     error
+		errorContains string
+		expectedToken string
+	}{
+		{
+			name:          "should return error when credentials are empty",
+			errorContains: "credentials are empty",
```

Comment:
This is for another iteration, but I think a better way is to assert on the error type, not on the error message. More work, but less error prone.

Reply:
Generally I agree, but it will require avoiding of simple error usage. Like .New(), fmt.Errorf(), etc.

Reply:
Only when you care about the error type and it affects further operations.